File: problems/linked-list/0206-reverse-linked-list.py
"""#206 Reverse Linked List
Invariant: previous is reversed; current begins the remaining unreversed suffix.
Time: O(n) | Space: O(1)
"""


class Solution:
    def reverseList(self, head):
        previous = None
        current = head

        while current is not None:
            following = current.next
            current.next = previous
            previous = current
            current = following

        return previous

File: problems/linked-list/0206-reverse-linked-list.js
/**
 * #206 Reverse Linked List
 * Invariant: previous is the reversed prefix; current begins the unreversed suffix.
 * Time: O(n) | Space: O(1)
 *
 * @param {ListNode|null} head
 * @return {ListNode|null}
 */
function reverseList(head) {
  let previous = null
  let current = head

  while (current !== null) {
    const next = current.next
    current.next = previous
    previous = current
    current = next
  }

  return previous
}

module.exports = reverseList
